### PR TITLE
C#: Fix GD0107 not applying to arrays and dictionaries containing nodes

### LINKED ISSUE #94589: `Godot.NET.Sdk`'s `GD0107` does not apply to exported members of type `Node[]`
### Tested versions

Reproducible in: 4.3.dev.mono [e25f3c0], 4.2.2.stable.mono.official

### System information

Windows 10 - Godot v4.2.2.stable.mono

### Issue description

`Godot.NET.Sdk` includes a diagnostic to prevent exports of nodes from resources. However, if you have an export of type `Node[]`, this diagnostic doesn't kick in and the project compiles. This results in the ability to create malformed resources that do not work correctly across scenes or after a reload of the editor.

### Steps to reproduce

- Download the MRP, see attachment below
- Open the scenes and MyResource.tres in a text editor. Observe that the `Res` field is defined for the nodes in the scenes, and the `Nodes` field is defined for the resources.
- Import the project into Godot, and build it. Observe that no diagnostics kick in.
- Observe that the resources have not loaded correctly.

### Minimal reproduction project (MRP)

[Csharp Export Diagnostics MRP.zip](https://github.com/user-attachments/files/16324976/Csharp.Export.Diagnostics.MRP.zip)

### PR DESCRIPTION
**Pull request by juanjp600**
Fixes #94589.

### CODE REVIEW

**Comment by raulsntos** [comment]
Thanks for contributing to the .NET module! This makes sense to me, and GDScript already reports an error for exported `Array[Node]` variables.

This PR currently only applies to `Node[]` but the diagnostic should also be reported for `Godot.Collections.Array<Node>`.

Also, since you are a new contributor, make sure to read [CONTRIBUTING.md](https://github.com/godotengine/godot/blob/4e5ed0bbfb56f0a71eb61c868f965476652c23df/CONTRIBUTING.md) and the [contributing documentation](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html) if you haven't already.

Feel free to reach out in the [development chat](https://chat.godotengine.org) if you need help.

**Comment by juanjp600** [comment]
Nice catch, thanks! I've added checks for `Godot.Collection.Array<T>` and `Godot.Collections.Dictionary<TKey, TValue>`.

Here's a new resource class to test with:
```cs
using Godot;
using System;

[GlobalClass]
public partial class AResource : Resource
{
	[Export] public ANode3D[] DotnetArrayField;
	[Export] public Godot.Collections.Array<ANode3D> GodotArrayField;
	[Export] public Godot.Collections.Dictionary<ANode3D, string> GodotDictionaryWithNodeAsKeyField;
	[Export] public Godot.Collections.Dictionary<string, ANode3D> GodotDictionaryWithNodeAsValueField;
	
	[Export] public ANode3D[] DotnetArrayProperty { get; set; }
	[Export] public Godot.Collections.Array<ANode3D> GodotArrayProperty { get; set; }
	[Export] public Godot.Collections.Dictionary<ANode3D, string> GodotDictionaryWithNodeAsKeyProperty { get; set; }
	[Export] public Godot.Collections.Dictionary<string, ANode3D> GodotDictionaryWithNodeAsValueProperty { get; set; }
}
```

**Comment by raulsntos** [comment]
You'll also need to add tests for the new cases that are expected to report the diagnostic in the [ExportDiagnostics_GD0107.cs](https://github.com/godotengine/godot/blob/4e5ed0bbfb56f0a71eb61c868f965476652c23df/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators.Tests/TestData/Sources/ExportDiagnostics_GD0107.cs) file.

The C# you just shared is pretty close, just need to add `{|GD0107:Identifier|}` in the places were the GD0107 diagnostic is expected.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators.Tests/TestData/Sources/ExportDiagnostics_GD0107.cs [hunk 1]
```diff
@@ -1,19 +1,68 @@
 using Godot;
+using Godot.Collections;
 
 public partial class ExportDiagnostics_GD0107_OK : Node
 {
     [Export]
     public Node NodeField;
 
+    [Export]
+    public Node[] DotnetArrayOfNodesField;
```
Comment (raulsntos) [inline modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators.Tests/TestData/Sources/ExportDiagnostics_GD0107.cs:10]: nit: We usually use the word `System` here.

```suggestion
    public Node[] SystemArrayOfNodesField;
```

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators.Tests/TestData/Sources/ExportDiagnostics_GD0107.cs [hunk 2]
```diff
@@ -1,19 +1,68 @@
...
+    public Dictionary<string, Node> GodotDictionaryWithNodeAsValueProperty { get; set; }
 }
 
 public partial class ExportDiagnostics_GD0107_KO : Resource
 {
     [Export]
     public Node {|GD0107:NodeField|};
 
+    [Export]
+    public Node[] {|GD0107:DotnetArrayOfNodesField|};
```
[the remark this answers is not in the document]
Comment (juanjp600) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators.Tests/TestData/Sources/ExportDiagnostics_GD0107.cs]: ...not sure how I missed that, will fix

**Review by raulsntos — APPROVED** [review]
Looks good to me, thanks!

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Comment by akien-mga** [comment]
Cherry-picked for 4.3.1.

**Pull request merged**